lib/Dialect/AIG/Analysis/LongestPathAnalysis.cpp

```diff
@@ -172,47 +262,1100 @@ struct llvm::DenseMapInfo<Object> {
+      continue;
+
+    for (auto *child : *node) {
+      auto childOp = child->getInstance();
+      if (!childOp || childOp->hasAttr("doNotPrint"))
+        continue;
+
+      worklist.push_back(child->getTarget());
+    }
+  }
+
+  // Initialize the local visitors if the module was visited, in the
+  // post-order of the instance graph.
+  for (auto module : topModules) {
+    auto *topNode = instanceGraph->lookup(module.getModuleNameAttr());
+    for (auto *node : llvm::post_order(topNode))
+      if (node && node->getModule())
+        if (auto hwMod = dyn_cast<hw::HWModuleOp>(*node->getModule())) {
+          if (visited.contains(hwMod))
+            ctx.localVisitors.insert(
+                {hwMod.getModuleNameAttr(),
+                 std::make_unique<LocalVisitor>(hwMod, &ctx)});
+        }
+
+    ctx.localVisitors[topNode->getModule().getModuleNameAttr()]->setTopLevel();
+  }
+
+  return mlir::failableParallelForEach(
+      module.getContext(), ctx.localVisitors,
+      [&](auto &it) { return it.second->initializeAndRun(); });
+}
+
+bool LongestPathAnalysis::Impl::isAnalysisAvailable(
+    StringAttr moduleName) const {
+  return ctx.localVisitors.find(moduleName) != ctx.localVisitors.end();
+}
+
+// Return the average of the maximum delays across all bits of the given
+// value, which is useful approximation for the delay of the value. For each
+// bit position, finds all paths and takes the maximum delay. Then averages
+// these maximum delays across all bits of the value.
+int64_t LongestPathAnalysis::Impl::getAverageMaxDelay(Value value) const {
+  SmallVector<DataflowPath> results;
+  size_t bitWidth = getBitWidth(value);
+  if (bitWidth == 0)
+    return 0;
+  int64_t totalDelay = 0;
+  for (size_t i = 0; i < bitWidth; ++i) {
+    // Clear results from previous iteration.
+    results.clear();
+    auto result = getResults(value, i, results);
+    if (failed(result))
+      return 0;
+
+    int64_t maxDelay = 0;
+    for (auto &path : results)
+      maxDelay = std::max(maxDelay, path.getDelay());
+    totalDelay += maxDelay;
+  }
+  return llvm::divideCeil(totalDelay, bitWidth);
+}
+
+int64_t LongestPathAnalysis::Impl::getMaxDelay(Value value) const {
+  SmallVector<DataflowPath> results;
+  size_t bitWidth = getBitWidth(value);
+  if (bitWidth == 0)
+    return 0;
+  int64_t maxDelay = 0;
+  for (size_t i = 0; i < bitWidth; ++i) {
+    // Clear results from previous iteration.
+    results.clear();
+    auto result = getResults(value, i, results);
+    if (failed(result))
+      return 0;
+
+    for (auto &path : results)
+      maxDelay = std::max(maxDelay, path.getDelay());
+  }
+  return maxDelay;
+}
+
 //===----------------------------------------------------------------------===//
 // LongestPathAnalysis
 //===----------------------------------------------------------------------===//
 
-LongestPathAnalysis::~LongestPathAnalysis() {}
+LongestPathAnalysis::~LongestPathAnalysis() { delete impl; }
```

Comment:
Could impl be an `std::unique_ptr` to avoid having to deal with naked pointers and allocation/deallocation?

Reply:
Yeah I tried but it didn't work. std::unique_ptr requires size to be determined in template instantiation so it's necessary to move Impl definition to the header. But we need to move implementation details such as Context/LocalVisitor so I think it's ok to use raw alloc/dealloc 😞